src/files/BrsFile.ts

```diff
@@ -1004,6 +1013,190 @@ export class BrsFile {
         return tokens[tokens.length - 1];
     }
 
+    /**
+     * Builds a list of document symbols for this file. Used by LanguageServer's onDocumentSymbol functionality
+     */
+    public async getDocumentSymbols() {
+        if (this.documentSymbols) {
+            return this.documentSymbols;
+        }
+
+        let symbols = [] as DocumentSymbol[];
+        await this.isReady();
+
+        for (const statement of this.ast.statements) {
+            const symbol = this.getDocumentSymbol(statement);
+            if (symbol) {
+                symbols.push(symbol);
+            }
+        }
+        this.documentSymbols = symbols;
+        return symbols;
+    }
+
+    /**
+     * Builds a list of workspace symbols for this file. Used by LanguageServer's onWorkspaceSymbol functionality
+     */
+    public async getWorkspaceSymbols() {
+        if (this.workspaceSymbols) {
+            return this.workspaceSymbols;
+        }
+
+        let symbols = [] as SymbolInformation[];
+        await this.isReady();
+
+        for (const statement of this.ast.statements) {
+            for (const symbol of this.generateWorkspaceSymbols(statement)) {
+                symbols.push(symbol);
+            }
+        }
+        this.workspaceSymbols = symbols;
+        return symbols;
+    }
+
+    /**
+     * Builds a single DocumentSymbol object for use by LanguageServer's onDocumentSymbol functionality
+     */
+    private getDocumentSymbol(statement: Statement) {
+        let symbolKind: SymbolKind;
+        let children = [] as DocumentSymbol[];
+        if (statement instanceof FunctionStatement) {
+            symbolKind = SymbolKind.Function;
+        } else if (statement instanceof NamespaceStatement) {
+            symbolKind = SymbolKind.Namespace;
+            for (const childStatement of statement.body.statements) {
+                const symbol = this.getDocumentSymbol(childStatement);
+                if (symbol) {
+                    children.push(symbol);
+                }
+            }
+        } else if (statement instanceof ClassStatement) {
+            symbolKind = SymbolKind.Class;
+            for (const childStatement of statement.body) {
+                const symbol = this.getDocumentSymbol(childStatement);
+                if (symbol) {
+                    children.push(symbol);
+                }
+            }
+        } else {
+            return;
+        }
+
+        const name = statement.getName(ParseMode.BrighterScript);
+        return DocumentSymbol.create(name, '', symbolKind, statement.range, statement.range, children);
+    }
+
+    /**
+     * Builds a single SymbolInformation object for use by LanguageServer's onWorkspaceSymbol functionality
+     */
+    private generateWorkspaceSymbols(statement: Statement, containerStatement?: ClassStatement | NamespaceStatement) {
+        let symbolKind: SymbolKind;
+        const symbols = [];
+
+        if (statement instanceof FunctionStatement) {
```

Comment:
Don't use instanceof for any of the AST traversal. Now that the plugin system exists, plugins can inject their own AST, which might not be the same binary instance. Prefer the `astUtils/reflection` functions `isFunctionStatement`, `isNamespaceStatement`, etc...)